File: solutions/sonnet-1/sonnet/perl/primes.pl
use strict;
use warnings;

my $n = <STDIN>;
$n = 0 unless defined $n;
$n =~ s/\D+//g;
$n = 0 if $n eq '';
$n = int($n);

my @is_composite = (0) x ($n + 1);
my @primes;

for (my $i = 2; $i <= $n; $i++) {
    next if $is_composite[$i];
    push @primes, $i;
    for (my $j = $i * $i; $j <= $n; $j += $i) {
        $is_composite[$j] = 1;
    }
}

print join(' ', @primes), "\n";

File: solutions/sonnet-1/sonnet/ruby/primes.rb
n = gets.to_i

if n < 2
  puts ""
else
  sieve = Array.new(n + 1, true)
  sieve[0] = false
  sieve[1] = false

  (2..Math.sqrt(n).to_i).each do |i|
    next unless sieve[i]
    (i * i).step(n, i) { |j| sieve[j] = false }
  end

  primes = (2..n).select { |i| sieve[i] }
  puts primes.join(" ")
end
